geoSegment.js

```diff
@@ -0,0 +1,353 @@
+        );
+      }
+    }
+
+    // Write SessionSeries to file
+    const sessionSeriesFilePath = getSessionSeriesFilePath(item.id);
+    await fs.writeJson(sessionSeriesFilePath, sessionSeriesData);
+    SESSION_SERIES_FILE_PATHS.push(sessionSeriesFilePath);
+
+    // Write ScheduledSessions to file
+    for (const generatedScheduledSession of scheduledSessionsForEveryEventSchedule) {
+      await linkScheduledSessionAndSessionSeriesAndWrite(generatedScheduledSession, sessionSeriesData);
+    }
+  });
+}
+
+/**
+ * @param {string} firehoseBaseUrl
+ * @param {string} firehoseApiKey
+ * @param {Segment[]} segments
+ */
+async function downloadSessionSeries(firehoseBaseUrl, firehoseApiKey, segments) {
+  console.log('downloadAndWriteSessionSeries() - starting');
+  // Get SessionSeries from Firehose
+  let nextUrl = `${firehoseBaseUrl}session-series`;
+  let backoffTimeInSeconds = RETRY_BACKOFF_MIN;
+  while (true) {
+    let nextNextUrl;
+    let items;
+    try {
+      ({ nextNextUrl, items } = await getFirehosePage(nextUrl, firehoseApiKey));
+    } catch (error) {
+      if (error instanceof Http404Error) {
+        console.error(`ERROR: URL ("${error.url}") not found. Please check the value for <rpde-endpoint>.`);
+        process.exit(1);
+      }
+      const loggableError = (error && error.isAxiosError)
+        ? loggableAxiosError(error)
+        : error;
+      console.warn(`WARN: Retrying [page: "${nextUrl}"] due to error:`, loggableError);
+      if (backoffTimeInSeconds > RETRY_BACKOFF_MAX) {
+        console.error('ERROR: Cannot download Firehose pages, backoff limit reached, terminating script');
+        process.exit(1);
+      }
+      await wait(backoffTimeInSeconds * 1000);
+      backoffTimeInSeconds *= RETRY_BACKOFF_EXPONENTIATION_RATE;
+      continue;
+    }
+
+    // Process SessionSeries items
+    await processSessionSeriesItems(items, segments);
+
+    // Have we reached the end of the feed?
+    if (nextNextUrl === nextUrl) {
+      break;
+    }
+
+    // If we haven't reached the end of the feed, carry on
+    nextUrl = nextNextUrl;
+    backoffTimeInSeconds = RETRY_BACKOFF_MIN;
+    await wait(MIN_REQUEST_DELAY_SECONDS * 1000);
+  }
+}
+
+/**
+ * @param {ScheduledSessionItem[]} items
+ */
+async function processScheduledSessionItems(items) {
+  items.forEach(async (scheduledSessionItem) => {
+    // Filter deleted ScheduledSessions
+    if (scheduledSessionItem.state === 'deleted') {
+      return;
+    }
+
+    // Filter ScheduledSessions in the past
+    if (!scheduledSessionItem.data.startDate || moment(scheduledSessionItem.data.startDate).isBefore(moment())) {
+      return;
+    }
+
+    // Does ScheduledSession have downloaded superEvent?
+    const scheduledSessionSuperEventId = scheduledSessionItem.data.superEvent;
+    const sessionSeriesFilePath = getSessionSeriesFilePath(scheduledSessionSuperEventId);
+    if (!SESSION_SERIES_FILE_PATHS.includes(sessionSeriesFilePath)) {
```

Comment:
As there's gonna be a lot of items in this array, perhaps this should be a set instead to reduce find time?

Reply:
Alternatively, since we're worried about robustness (e.g. we don't know how large this array/set could grow), we could just use a file system call to check if the SessionSeries exists. Slower, but guarantueed small memory footprint

It looks like fs.readJson supports just returning null if the item isn't there (https://github.com/jprichardson/node-fs-extra/blob/master/docs/readJson.md)

This does seem in the spirit of the Trello card's comments about minimal memory usage

Reply:
+1 to filesystem check assuming performance isn't heavily impacted. As we're already writing a file for each record this should just 2X the run time (rather than order of magnitude) assuming read time is similar to write.

Reply:
It's also occurred to me that there's a risk they'll run out of inodes on the linux filesystem as we're writing one record for every ScheduledSession, but that's probably just a risk we flag to them

Reply:
Have changed to readJson(path, {throws:false})

Reply:
>  It's also occurred to me that there's a risk they'll run out of inodes

@nickevansuk that's a good point. Have we already talked to them about this limitation in the past..? It's ringing bells. And not unsurprising given how many files are being created!